xds/internal/xdsclient/transport/lrs/lrs_stream.go

```diff
@@ -0,0 +1,326 @@
+		defer lrs.mu.Unlock()
+
+		if lrs.refCount == 0 {
+			lrs.logger.Errorf("Attempting to stop an LRS stream that has already been stopped.")
+			return
+		}
+		lrs.refCount--
+		if lrs.refCount != 0 {
+			return
+		}
+		lrs.cancelStream()
+		lrs.cancelStream = nil
+		lrs.logger.Infof("Stopping LRS stream")
+	})
+
+	if lrs.refCount != 0 {
+		lrs.refCount++
+		return lrs.lrsStore, cleanup
+	}
+
+	lrs.refCount++
+	ctx, cancel := context.WithCancel(context.Background())
+	lrs.cancelStream = cancel
+	lrs.doneCh = make(chan struct{})
+	go lrs.runner(ctx)
+	return lrs.lrsStore, cleanup
+}
+
+// runner is responsible for managing the lifetime of an LRS streaming call. It
+// creates the stream, sends the initial LoadStatsRequest, receives the first
+// LoadStatsResponse, and then starts a goroutine to periodically send
+// LoadStatsRequests. The runner will restart the stream if it encounters any
+// errors.
+func (lrs *Stream) runner(ctx context.Context) {
+	defer close(lrs.doneCh)
+
+	// This feature indicates that the client supports the
+	// LoadStatsResponse.send_all_clusters field in the LRS response.
+	node := proto.Clone(lrs.nodeProto).(*v3corepb.Node)
+	node.ClientFeatures = append(node.ClientFeatures, "envoy.lrs.supports_send_all_clusters")
+
+	runLoadReportStream := func() error {
+		// streamCtx is created and canceled in case we terminate the stream
+		// early for any reason, to avoid gRPC-Go leaking the RPC's monitoring
+		// goroutine.
+		streamCtx, cancel := context.WithCancel(ctx)
+		defer cancel()
+
+		stream, err := lrs.transport.CreateStreamingCall(ctx, "/envoy.service.load_stats.v3.LoadReportingService/StreamLoadStats")
+		if err != nil {
+			lrs.logger.Warningf("Creating new LRS stream failed: %v", err)
+			return nil
+		}
+		if lrs.logger.V(2) {
+			lrs.logger.Infof("LRS stream created")
+		}
+
+		if err := lrs.sendFirstLoadStatsRequest(stream, node); err != nil {
+			lrs.logger.Warningf("Sending first LRS request failed: %v", err)
+			return nil
+		}
+
+		clusters, interval, err := lrs.recvFirstLoadStatsResponse(stream)
+		if err != nil {
+			lrs.logger.Warningf("Reading from LRS stream failed: %v", err)
+			return nil
+		}
+
+		// We reset backoff state when we successfully receive at least one
+		// message from the server.
+		lrs.sendLoads(streamCtx, stream, clusters, interval)
+		return backoff.ErrResetBackoff
+	}
+	backoff.RunF(ctx, runLoadReportStream, lrs.backoff)
+}
+
+func (lrs *Stream) sendLoads(ctx context.Context, stream transport.StreamingCall, clusterNames []string, interval time.Duration) {
+	tick := time.NewTicker(interval)
+	defer tick.Stop()
+	for {
+		select {
+		case <-tick.C:
+		case <-ctx.Done():
+			return
+		}
+		if err := lrs.sendLoadStatsRequest(stream, lrs.lrsStore.Stats(clusterNames)); err != nil {
+			lrs.logger.Warningf("Writing to LRS stream failed: %v", err)
+			return
+		}
+	}
+}
+
+func (lrs *Stream) sendFirstLoadStatsRequest(stream transport.StreamingCall, node *v3corepb.Node) error {
```

Comment:
why first send and first receive have different implementation?

Reply:
The way LRS works is as follows:
- The client sends a first request which only contains the Node proto that identifies the client instance
- It waits for the server to send a response which contains two things
  - The list of clusters for which the server is interested in receiving load reports
  - The load reporting interval
- The client does not read any more messages from the server after this point
- For every load reporting interval, the client sends a bunch of cluster load reports

The first message and the subsequent messages being sent use the same proto message `LoadStatsRequest`. Hence we need different implementations for them.

As far as the receive side goes, there is only one read that ever happens. So, we *could* have not called it `recvFirstLoadStatsResponse` and we *could* have called it simply `recvLoadStatsResponse`. But I feel the former is more descriptive.

Hope this helps.